### DISCUSSION TITLE
Feature request: make consult-line treat initial empty input as symbol at point (similar to helm-occur)

### DISCUSSION THREAD

**Discussion by mpedramfar** [minad/consult#848] [Ideas]
Hi and thanks for the great package!

Helm has a variable called `helm-maybe-use-default-as-input` and a list `helm-sources-using-default-as-input`. When `helm-maybe-use-default-as-input` is non-nil or the function that's being called, i.e. `helm-occur`, belongs to `helm-sources-using-default-as-input`, helm treats empty input as the symbol at point. More specifically, if the point is at the word "hello" and I run `helm-occur`, it will match all of the instances of the word "hello" in the current buffer. The moment I input a single character, it will match my input instead. 

Is there a way to make `consult-line` behave similarly? Currently, when `consult-line` is called, while input is still empty, it will match all non-empty lines in the buffer. I don't think matching all non-empty lines in the current buffer is common use case.

There is a partial solution in [the wiki](https://github.com/minad/consult/wiki#start-consult-line-search-with-symbol-at-point). However, it doesn't treat empty input as symbol at point, it fills the input with symbol at point. So using this means that, if I want to search for something other than the symbol at point, I have to either move the point to an empty area, or call this function and clear the minibuffer before I can start my search.

**Comment by minad** [comment]
Thanks. That's a nice idea. Right now you can use the future history to obtain the symbol at point. I've not considered such prefilled input where the input is going away as soon as you type. This is certainly doable but differs quite a bit from the usual built-in completion command behavior and I try to avoid such deviations. Nevertheless it would be interesting to try this out.

**Comment by mpedramfar** [comment]
I looked a bit and I agree that doing it through built-in `completing-read` doesn't seem straightforward. But we don't actually need to tweak that, we can just advice our completion functions:

```emacs-lisp
(setq orderless-empty-input-is-matched-with "")

(defun my-completion-advice (fun string table pred _point)
  (let ((string (or (and (equal string "")
                         orderless-empty-input-is-matched-with)
                    string)))
    (funcall fun string table pred _point)))

(advice-add #'orderless-try-completion  :around #'my-completion-advice)
(advice-add #'orderless-all-completions :around #'my-completion-advice)

(defun my-consult-line ()
  (interactive)
  (let ((orderless-empty-input-is-matched-with (thing-at-point 'symbol)))
    (consult-line)))
```

**Reply by minad** [reply]
Nice. To clarify - I didn't mean that it isn't straightforward. There are multiple ways to implement this feature, one via adjusting the completion style as you demonstrated. One could also insert text with an overlay in the minibuffer. The overlay could have a modification hook, which deletes the overlay. My point is really only that the behavior differs from the usual Emacs ways.

**Reply by mpedramfar** [reply]
I see, your solution with overlays seems to be cleaner (i.e. not dependent on the choice of completion style). Would you consider adding this option through a customizable variable with the default being the current behaviour?

**Reply by minad** [reply]
No, I prefer to not add this since M-n M-n doesn't seem too bad to me. Also I am not fond of deviating too much from existing commands. We could document this in the wiki though. Maybe it is possible to find a unified solution which would work for all commands then we could consider adding it to Vertico.

**Reply by mpedramfar** [reply]
Yeah, it is a bit of a deviation. Maybe vertico is a better place for something like this. Something general, like a variable `vertico-empty-input-is-matched-with` that can be a string or a function/form that returns a string and every other non-nil value means symbol-at-point.

The reason I'm mentioning about the option of function/form is that maybe we're at an org-link and the command we're calling works with links instead of the description of the links, so `(thing-at-point 'symbol)` is not the universal solution for what should be the default for every possible command.

**Reply by mpedramfar** [reply]
We could also have the variable `vertico-empty-input-is-matched-with` integrate with embark, e.g. setting it to the symbol `'embark` is equivalent to the form `(plist-get (car (embark--targets)) :target)`.

**Reply by mpedramfar** [reply]
I'm not sure what I said about the variable being a function/form is actually useful. Here is a version where `vertico-default-input` can be either a string or nil.
```emacs-lisp
(setq vertico-default-input nil)

;; might be best to change the value of `content' within `vertico--update' instead
;; but I can't advice `vertico--update' this easily
(defun my-vertico--recompute-advice (fun pt content)
  (let* ((content (or (and (equal vertico--input t) vertico-default-input)
                      content)))
    (funcall fun pt content)))

(advice-add #'vertico--recompute :around #'my-vertico--recompute-advice)

(defun my-consult-line ()
  (interactive)
  (let ((vertico-default-input (thing-at-point 'symbol)))
    (consult-line)))
```

**Comment by oantolin** [comment]
This seems like a lot of effort just to avoid pressing `M-n`. 🤷🏽

**Reply by mpedramfar** [reply]
It's two extra keystrokes, and spending a lot of time and effort just to make the workflow a tiny bit more "efficient" is the emacs way after all! :smile: 

Besides, I'm not sure why `M-n` doesn't work for me. After I call `consult-line`, it moves the point to the beginning of the first non-empty line after point and when I press `M-n`, it selects the whole line, not just the symbol at point.

**Reply by minad** [reply]
I tend to agree with @oantolin here. While it is certainly nice to optimize workflows I also think one should weigh efficiency gains, implementation efforts and accumulated complexity.

**Reply by mpedramfar** [reply]
Fair enough. I already have the functionality I want with the code snippet above.

**Reply by oantolin** [reply]
I made a mistake, @mpedramfar: you need to press `M-n` *twice* for the symbol at point. This makes me more sympathetic to your plight. :) To be honest I never use `M-n M-n` in `consult-line`, what I do instead is use embark to start consult-line with the symbol at point: `C-. C l` (`C-.` is my binding for `embark-act`, `C` is the prefix for Consult, and `l` for `consult-line`).

**Reply by minad** [reply]
Right, using Embark for this is so much better. There are already multiple simple ways to achieve the desired behavior in some form or another.

- Via Embark (C-c . C l)
- Future history (yes, it needs M-n twice)
- Setting :initial via consult-customize, maybe also via a separate comand consult-line-symbol-at-point 

Given these alternatives I don't think we need yet another one. Note that one can also override the future history such that the default values are removed (see `minibuffer-default-add-function`). Then a single M-n will suffice. The future history contains the default(s), additional strings and all the candidate strings from the table. I find this design questionable (in particular, why do I need all the values?) and has also been criticized on emacs-devel, in particular by the Icicles author.

**Reply by mpedramfar** [reply]
Oh, I didn't realize I would need to press it twice. Maybe the README could be improved.

As for using embark, I just tried it and it selected the whole line! So I ended up having to press `C-. C l M-n M-n` to select the symbol at point!

**Reply by minad** [reply]
> Oh, I didn't realize I would need to press it twice. Maybe the README could be improved.

Sure, everything can be improved. The Emacs manual also documents the future history: https://www.gnu.org/software/emacs/manual/html_node/emacs/Minibuffer-History.html.

> As for using embark, I just tried it and it selected the whole line!

This can happen if the target finder for the line/paragraph has a high priority (comes early in the list). Then you can cycle by pressing C-. again. But I also think that some of the target finders in Embark should be reordered, such that the smaller targets are prioritized higher more often.

**Reply by oantolin** [reply]
Embark, in it's default configuration, almost always selects the symbol at point first. There are exceptions, but not that many. For example in Org  mode buffers it gives preference to org targets, such as the current list item. What major mode was the buffer you were in, @mpedramfar?

**Reply by minad** [reply]
> For example in Org mode buffers it gives preference to org targets, such as the current list item.

This is what I had just observed, where the org list item got selected before smaller targets.

**Reply by oantolin** [reply]
Target ordering is super hard. I sometimes kind of want different orders for embark-dwim and embark-act. It would be nice if embark-dwim toggled the org list item's checkbox, but embark-act C l ran consult-line on the symbol at point...

**Reply by minad** [reply]
Yes, I agree. There are various possibilities which might help. 

For example one could restrict the region of targets in the style of the Org speed keys, e.g., the list items would only match when point is located on top of the `-` or the `- [ ]`. Then one could add another lower priority target finder which matches the item everywhere and which is accessible via cycling.

Another perhaps simpler option is to move all all the finders which are always small or highly restricted to the front. In my setup the list looks like this (with some custom additions):

(embark-target-jinx-at-point embark-target-todo-at-point embark-target-pkg-at-point embark--vertico-selected embark-target-top-minibuffer-completion embark-target-active-region embark-org-target-element-context embark-org-target-agenda-item embark-target-text-heading-at-point embark-target-collect-candidate embark-target-completion-at-point embark-target-bug-reference-at-point embark-target-flymake-at-point embark-target-smerge-at-point embark-target-package-at-point **embark-target-email-at-point** **embark-target-url-at-point** embark-target-file-at-point **embark-target-custom-variable-at-point** **embark-target-identifier-at-point** embark-target-library-file-at-point embark-target-expression-at-point embark-target-sentence-at-point embark-target-paragraph-at-point embark-target-defun-at-point embark-target-prog-heading-at-point)

The ones I marked bold should come earlier, I think, before embark-org-target-element at least.

**Reply by mpedramfar** [reply]
I see. I was in org-mode at a heading and the whole heading got selected.

**Reply by oantolin** [reply]
Well, if it's not the first choice, you can still target-cycle until you get to the symbol at point. In my setup it's the second target, right after the heading, so `C-. C-. C l` would work, @mpedramfar.

Also, @minad, what about just moving `embark-org-target-element-context` and `embark-target-text-heading-at-point` to right before `embark-target-expression-at-point`? Oh, no, that's not enough: I'd probably then have to tweak `embark-target-identifier-at-point` so symbols like `*`, `**` and `-` are not recognized in org mode so that at least at the start of the heading you get the heading target first...

**Reply by minad** [reply]
@oantolin I am really unsure. Ideally we avoid such inter-finder dependencies and ad-hoc checks.

**Reply by oantolin** [reply]
Yeah, this is hard to get right. :(

**Reply by oantolin** [reply]
Maybe symbols without any letters should only be recognized in prog modes.

**Reply by mpedramfar** [reply]
Yeah, I'm using the default order of embark targets and the symbol at point is the second target (in an org-mode heading).
It would've been better, for `consult-line`, if it was the first, but I'm still new to embark and I don't have strong opinions about the ordering of targets.

**Comment by Niluge-KiWi** [comment]
> Right, using Embark for this is so much better. There are already multiple simple ways to achieve the desired behavior in some form or another.
> 
>     * Via Embark (C-c . C l)
> 
>     * Future history (yes, it needs M-n twice)
> 
>     * Setting :initial via consult-customize, maybe also via a separate comand consult-line-symbol-at-point
> 
> 
> Given these alternatives I don't think we need yet another one. Note that one can also override the future history such that the default values are removed (see `minibuffer-default-add-function`). Then a single M-n will suffice. The future history contains the default(s), additional strings and all the candidate strings from the table. I find this design questionable (in particular, why do I need all the values?) and has also been criticized on emacs-devel, in particular by the Icicles author.


(new user of consult/vertico/etc here, having declared .emacs bankruptcy after 15 years of cruft accumulation and stagnation)
I'm tempted to try the "future history" way (notably because I was first pointed-to-it by the readme here at at least 2 locations: it seems nice and short), but I didn't succeed following your hints @minad :

what I understood after some digging:
- first M-n calls `goto-history-element` with narg=-1
- (in my emacs 29.1) in such case it skips calling `(funcall minibuffer-default-add-function)` to override `minibuffer-default`
- so it uses `minibuffer-default`, the current line when called in `consult-line`
=> I have no idea how to "override the future history such that the default values are removed" by working with `minibuffer-default-add-function`

(`minibuffer-default-add-function` *is* called in second M-n, defined in `consult--read-1`, but that is too late)

**Reply by Niluge-KiWi** [reply]
After reading https://github.com/minad/consult/issues/417 I understand that trying to use `consult-line` to replace `isearch` doesn't really make sense, so I stopped trying that direction, and my message above doesn't really make sense anymore.

(I've open a discussion about that in emacs-bedrock, where I come from and got that idea initially)
